rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,833 @@
+  to find an optimal grouping of TE sub-expressions, ii) an analytic model of cost to guide the search,
+  and iii) cascading scheduling of the TE sub-expressions so as to reduce memory high-watermark. By contrast
+  Collage i) also uses dynamic-programming, but to find an optimal grouping of Relay sub-expressions, ii)
+  uses measurement to guide the search and iii) assuming the toolchain will 'do its best' with the
+  sub-graph offered to it.
+- The [Universal modular Accelerator Interface](https://github.com/apache/tvm-rfcs/pull/60) proposal
+  adds a layer on top of the existing and separate TVM BYOC, operator strategy, operator scheduling,
+  target-specific passes and target-specific code generation extension points. Collage currently relies
+  only on the global pattern registry and global `relay.ext.<toolchain>` function to integrate with BYOC
+  integrations, but this is trivial to change should this project change the source of truth.
+
+## Example
+
+We start with `mod` bound to [MNIST](https://github.com/onnx/models/tree/main/vision/classification/mnist):
+
+```
+fn (%x: Tensor[(1, 1, 28, 28), float32]) -> Tensor[(1, 10), float32] {
+  %0 = nn.pad(%x, 0f, pad_width=[[0, 0], [0, 0], [2, 2], [2, 2]]);
+  %1 = nn.conv2d(%0, meta[relay.Constant][0] /*Tensor[(8, 1, 5, 5), float32]*/,
+                 padding=[0, 0, 0, 0], channels=8, kernel_size=[5, 5]);
+  %2 = add(%1, meta[relay.Constant][1] /*Tensor[(8, 1, 1), float32]*/);
+  %3 = nn.relu(%2);
+  %4 = nn.max_pool2d(%3, pool_size=[2, 2], strides=[2, 2], padding=[0, 0, 0, 0]);
+  %5 = nn.pad(%4, 0f, pad_width=[[0, 0], [0, 0], [2, 2], [2, 2]]);
+  %6 = nn.conv2d(%5, meta[relay.Constant][2] /*Tensor[(16, 8, 5, 5), float32]*/,
+                 padding=[0, 0, 0, 0], channels=16, kernel_size=[5, 5]);
+  %7 = add(%6, meta[relay.Constant][3] /*Tensor[(16, 1, 1), float32]*/);
+  %8 = nn.relu(%7);
+  %9 = nn.max_pool2d(%8, pool_size=[3, 3], strides=[3, 3], padding=[0, 0, 0, 0]);
+  %10 = reshape(%9, newshape=[1, 256]);
+  %11 = nn.dense(%10, meta[relay.Constant][4] /*Tensor[(10, 256), float32]*/, units=None, out_dtype="float32");
+  add(%11, meta[relay.Constant][5] /*Tensor[(1, 10), float32]*/)
+}
+```
+
+We can compile this with Collage enabled for a variety of NVIDIA toolchains/libraries as follows:
+
+```
+with tvm.transform.PassContext(config={"relay.fallback_device_type": 2, "relay.collage.enable_collage": True}):
+    host_target = tvm.target.Target("llvm")
+    generic_target = tvm.target.Target("cuda", host_target)
+    cutlass_target = tvm.target.Target("cuda -compiler=cutlass", host_target)
+    tensorrt_target = tvm.target.Target("cuda -compiler=tensorrt", host_target)
+    cudnn_target = tvm.target.Target("cuda -libs=cudnn", host_target)
+    cublas_target = tvm.target.Target("cuda -libs=cublas", host_target)
+    targets = [generic_target, cutlass_target, tensorrt_target, cudnn_target, cublas_target]
+    exe = tvm.relay.vm.compile(mod, target=targets)
```

Comment:
The targets list here is a kind of `composite` target.  Can we just use it?
https://github.com/apache/tvm/blob/0b6125637058dd080371399e913314bc5f132af3/src/target/target_kind.cc#L406-L407

Reply:
Thanks. Ideally I'd not need to innovate in target handling so if there's an easier path I'm all for it. I do see https://discuss.tvm.apache.org/t/rfc-composite-target/7744/10 which explains the intent is similar to what I need here, but I don't see any downstream handling of it -- do you have any pointers? 

There's also been quite a bit of discussion lately about handling multiple targets which I've not followed, perhaps there's a summary of that?

Reply:
As far as I know, there is no such handling for the `composite` target yet. I think we do not have a summary either.

Reply:
Marking as 'resolved', and added Appendix G to give more detail.